syntax = "proto3";

import "types.proto";

message RegisterWorkerRequest {
  string address = 1;
}

message RegisterWorkerReply {
  uint64 workerid = 1;
}

message RemoteCallRequest {
  string name = 1;
  Values arg = 2;
}

message RemoteCallReply {
  uint64 result = 1;
}

message PullObjectRequest {
  uint64 ref = 1;
}

service Orchestra {
  rpc RegisterWorker(RegisterWorkerRequest) returns (RegisterWorkerReply);
  // rpc RegisterFunction
  rpc RemoteCall(RemoteCallRequest) returns (RemoteCallReply);
  // rpc PushObject
  // rpc PullObject(PullObjectRequest)
}

message InvokeCallRequest {

}

message InvokeCallReply {

}

service Worker {
  rpc InvokeCall(InvokeCallRequest) returns (InvokeCallReply);
  // rpc PushObj(PushObjRequest) returns (PushObjReply);
  // rpc RequestTransfer(RequestTransferRequest) returns (RequestTransferReply);
}
